README.md
# rs-sqlite

Learn [Let's Build a Simple Database](https://cstack.github.io/db_tutorial/) with Rust.

CHANGELOG.md
第八章开始, 对C语言有点不是很理解, 先暂停

spec/main_spec.rb
describe "database" do
    before do
        `cargo b -r`
        `rm -rf test.db`
    end
    
    after(:all) do
        `rm -rf test.db`
    end

    def run_script(commands)
        raw_output = nil
        IO.popen("./target/release/rs-sqlite test.db", "r+") do |pipe|
            commands.each do |command|
                begin
                    pipe.puts command
                rescue Errno::EPIPE
                    break
                end
        end
            pipe.close_write
        
            # Read entire output
            raw_output = pipe.gets(nil)
        end
        raw_output.split("\n")
    end
    
    it 'inserts and retrieves a row' do
        result = run_script([
            "insert 1 user1 person1@example.com",
            "select",
            ".exit",
        ])
        # match_array 是没有检测顺序的
        expect(result).to eq([
            "db > Executed.",
            "db > (1, user1, person1@example.com)",
            "Executed.",
            "db > Bye~", 
        ])
    end

    it 'prints error message when table is full' do
        script = (1..1401).map do |i|
            "insert #{i} user#{i} person#{i}@example.com"
        end
        script << ".exit"
        result = run_script(script)
        expect(result[-2]).to eq('db > Error: Table full.')
    end

    it 'allows inserting strings that are the maximum length' do
        long_username = "a" * 32
        long_email = "a" * 255
        script = [
            "insert 1 #{long_username} #{long_email}",
            "select",
            ".exit",
        ]
        result = run_script(script)
        expect(result).to eq([
            "db > Executed.",
            "db > (1, #{long_username}, #{long_email})",
            "Executed.",
            "db > Bye~",
        ])
    end
    
    it 'prints error message if strings are too long' do
        long_username = "a" * 33
        long_email = "a" * 256
        script = [
            "insert 1 #{long_username} #{long_email}",
            "select",
            ".exit",
        ]
        result = run_script(script)
        expect(result).to eq([
            "db > String is too long.",
            "db > Executed.",
            "db > Bye~",
        ])
    end
    
    it 'prints an error message if id is negative' do
        script = [
            "insert -1 startdusk foo@bar.com",
            "select",
            ".exit",
        ]
        result = run_script(script)
        expect(result).to eq([
            "db > ID must be positive.",
            "db > Executed.",
            "db > Bye~",
        ])
    end

    it 'keeps data after closing connection' do
        result1 = run_script([
            "insert 1 user1 person1@example.com",
            ".exit",
        ])
        expect(result1).to eq([
            "db > Executed.",
            "db > Bye~",
        ])
        result2 = run_script([
            "select",
            ".exit",
        ])
        expect(result2).to eq([
            "db > (1, user1, person1@example.com)",
            "Executed.",
            "db > Bye~",
        ])
    end

end


tests/integration_test.rs
#[cfg(test)]
mod tests {
    use std::{fs, path::Path};

    struct DbFile {
        name: String,
    }

    impl DbFile {
        pub fn clean(db_name: &str) -> Self {
            if Path::new(db_name).exists() {
                fs::remove_file(db_name).unwrap();
            }
            Self {
                name: db_name.to_string(),
            }
        }
    }

    impl Drop for DbFile {
        fn drop(&mut self) {
            fs::remove_file(&self.name).unwrap();
        }
    }

    #[test]
    fn main_test() {
        use assert_cmd::Command;

        let db_name = "test.db";
        // 注意: 一定要提取变量( _ 只是忽略返回值), 才会执行 Drop trait
        let _db_file = DbFile::clean(db_name);

        // first create database
        let bin_name = env!("CARGO_PKG_NAME");
        let mut cmd = Command::cargo_bin(bin_name).unwrap();
        let session = cmd.arg(db_name);
        session.assert().stdout("db > ");
        session
            .write_stdin("INSERT 1 abc abc@gmail.com")
            .assert()
            .stdout("db > Executed.\ndb > ");

        session
            .write_stdin("SELECT")
            .assert()
            .stdout("db > (1, abc, abc@gmail.com)\nExecuted.\ndb > ");
        session
            .write_stdin("INSERT 2 byd byd@gmail.com")
            .assert()
            .stdout("db > Executed.\ndb > ");
        session
            .write_stdin("SELECT")
            .assert()
            .stdout("db > (1, abc, abc@gmail.com)\n(2, byd, byd@gmail.com)\nExecuted.\ndb > ");
        session.write_stdin(".exit").assert().stdout("db > Bye~\n");

        // test database persistence
        let mut cmd = Command::cargo_bin(bin_name).unwrap();
        let session = cmd.arg(db_name);
        session.assert().stdout("db > ");
        session
            .write_stdin("SELECT")
            .assert()
            .stdout("db > (1, abc, abc@gmail.com)\n(2, byd, byd@gmail.com)\nExecuted.\ndb > ");
        session
            .write_stdin("INSERT 3 li li@gmail.com")
            .assert()
            .stdout("db > Executed.\ndb > ");
        session
            .write_stdin("SELECT")
            .assert().stdout("db > (1, abc, abc@gmail.com)\n(2, byd, byd@gmail.com)\n(3, li, li@gmail.com)\nExecuted.\ndb > ");
        session.write_stdin(".exit").assert().stdout("db > Bye~\n");
    }
}


src/lib.rs
use binary_serde::{BinarySerde, DeserializeError, Endianness};

pub const EXIT_SUCCESS: i32 = 0;
pub const EXIT_FAILURE: i32 = -1;
pub const PAGE_SIZE: usize = 4096;

// windows 的栈要比 unix 的要小, 数组是在栈上分配, 改为vec在堆上分配
pub const TABLE_MAX_PAGES: usize = 100;

// ROW_SIZE = 291
pub const ROW_SIZE: usize = <Row as BinarySerde>::MAX_SERIALIZED_SIZE;
// ROWS_PER_PAGE 每个页到底有多少行, 291是计算Row结构体的所有字段的总长度
pub const ROWS_PER_PAGE: usize = PAGE_SIZE / ROW_SIZE;
// TABLE_MAX_ROWS 表最大行数
pub const TABLE_MAX_ROWS: usize = ROWS_PER_PAGE * TABLE_MAX_PAGES;

pub type RowLine = [u8; ROW_SIZE];
pub type Page = [u8; PAGE_SIZE];

#[repr(C)]
#[derive(Debug, BinarySerde)]
pub struct Row {
    pub id: u32,
    pub username: Username,
    pub email: Email,
}

impl Row {
    pub fn encode(&self) -> RowLine {
        let mut insert_data: RowLine = [0u8; ROW_SIZE];
        self.binary_serialize(&mut insert_data, Endianness::Little);
        insert_data
    }

    pub fn from(buf: &[u8]) -> Result<Row, DeserializeError> {
        Row::binary_deserialize(buf, Endianness::Little)
    }
}

pub type Username = [u8; 32];

pub type Email = [u8; 255];

impl std::fmt::Display for Row {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let username = String::from_utf8(self.username.to_vec())
            .unwrap()
            .replace('\0', "");
        let email = String::from_utf8(self.email.to_vec())
            .unwrap()
            .replace('\0', "");
        write!(f, "({}, {}, {})", self.id, username, email)
    }
}

pub fn to_u8_array<const N: usize>(s: &str) -> [u8; N] {
    let mut a = [0u8; N];
    s.bytes().zip(a.iter_mut()).for_each(|(b, ptr)| *ptr = b);
    a
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn to_u8_array_should_work() {
        assert_eq!(to_u8_array::<10>("0000000000"), [48u8; 10])
    }
}


src/main.rs
use std::env::args;
use std::io::{stdin, stdout, BufRead, Write};
use std::process::exit;

mod pager;
mod table;
use table::Table;

use rs_sqlite::EXIT_FAILURE;

fn main() {
    let args: Vec<String> = args().collect();
    if args.len() < 2 {
        println!("Must supply a database filename.");
        exit(EXIT_FAILURE);
    }

    let filename = &args[1];

    let stdin = stdin();
    print_prompt();
    let _ = stdout().flush();
    let mut table = Table::db_open(filename);
    for line in stdin.lock().lines() {
        let Ok(line) = line else {
            break;
        };
        if line.is_empty() {
            break;
        }

        table.run(&line.trim().to_lowercase());

        print_prompt();
        let _ = stdout().flush();
    }
}

fn print_prompt() {
    print!("db > ");
}


src/pager.rs
use std::{fs::File, process::exit};

#[cfg(target_family = "unix")]
use std::os::unix::prelude::FileExt;
#[cfg(target_family = "windows")]
use std::os::windows::prelude::FileExt;

use rs_sqlite::{Page, RowLine, EXIT_FAILURE, PAGE_SIZE, ROWS_PER_PAGE, ROW_SIZE, TABLE_MAX_PAGES};

pub struct Pager {
    pub file_descriptor: File,
    pub file_length: usize,
    pub pages: Vec<Option<Vec<u8>>>,
}

impl Pager {
    pub fn open(filename: &str) -> std::io::Result<Self> {
        let file = File::options()
            .write(true)
            .create(true)
            .append(true)
            .read(true)
            .open(filename)?;

        let file_length = file.metadata().unwrap().len() as usize;
        let pages = vec![None; TABLE_MAX_PAGES];
        Ok(Self {
            file_descriptor: file,
            file_length,
            pages,
        })
    }

    pub fn save_row(&mut self, row_num: usize, row: RowLine) {
        let page_num = self.page_num(row_num);
        if page_num > TABLE_MAX_PAGES {
            println!(
                "Tried to fetch page number out of bounds. {} > {}",
                page_num, TABLE_MAX_PAGES
            );
            exit(EXIT_FAILURE);
        }

        let offset = self.offset(row_num);

        #[cfg(target_family = "unix")]
        self.file_descriptor.write_at(&row, offset as u64).unwrap();

        #[cfg(target_family = "windows")]
        self.file_descriptor
            .seek_write(&row, offset as u64)
            .unwrap();

        let mut page = self.get_page(row_num);
        page[offset..(row.len() + offset)].copy_from_slice(&row[..]);
        self.pages[page_num] = Some(page.to_vec());
    }

    pub fn get_page(&mut self, row_num: usize) -> Page {
        let page_num = self.page_num(row_num);
        if page_num > TABLE_MAX_PAGES {
            println!(
                "Tried to fetch page number out of bounds. {} > {}",
                page_num, TABLE_MAX_PAGES
            );
            exit(EXIT_FAILURE);
        }
        let mut tmp_page: Page = [0u8; PAGE_SIZE];
        let Some(page) = &self.pages[page_num] else {
            // Cache miss. Allocate memory and load from file.
            let mut num_pages = self.file_length / PAGE_SIZE;

            // We might save a partial page at the end of the file
            if (self.file_length % PAGE_SIZE) > 0 {
                num_pages += 1;
            }
            // 读取指定数据
            if num_pages > 0 && page_num <= num_pages {
                let offset = (page_num * PAGE_SIZE) as u64;
                #[cfg(target_family = "unix")]
                self.file_descriptor.read_at(&mut tmp_page,  offset).unwrap();
                #[cfg(target_family = "windows")]
                self.file_descriptor.seek_read(&mut tmp_page,  offset).unwrap();
            }
            self.pages[page_num] = Some(tmp_page.to_vec());
            return tmp_page
        };

        tmp_page[..page.len()].copy_from_slice(&page[..]);
        tmp_page
    }

    pub fn get_row(&mut self, row_num: usize) -> RowLine {
        let page = self.get_page(row_num);
        let start = self.offset(row_num);
        let end = start + ROW_SIZE;
        let select_data = page[start..end].to_vec();
        let data_length = select_data.len();
        let array: RowLine = match select_data.try_into() {
            Ok(ba) => ba,
            Err(_) => panic!(
                "Expected a row of length {} but it was {}",
                ROW_SIZE, data_length,
            ),
        };
        array
    }

    pub fn free(&mut self) {
        for i in 0..self.pages.len() {
            self.pages[i] = None;
        }
    }

    fn page_num(&self, row_num: usize) -> usize {
        row_num / ROWS_PER_PAGE
    }

    fn offset(&self, row_num: usize) -> usize {
        let row_offset = row_num % ROWS_PER_PAGE;
        row_offset * ROW_SIZE
    }
}


src/cursor.rs
use rs_sqlite::{ROWS_PER_PAGE, ROW_SIZE};

pub struct Cursor {
    row_num: usize,
    end_of_table: bool, // Indicates a position one past the last element
}

impl Cursor {
    pub fn table_start(row_num: usize) -> Self {
        Self {
            row_num: 0,
            end_of_table: row_num == 0,
        }
    }

    pub fn table_end(row_num: usize) -> Self {
        Self {
            row_num,
            end_of_table: true,
        }
    }

    pub fn offset(&self) -> usize {
        let row_num = self.row_num;
        let row_offset = row_num % ROWS_PER_PAGE;
        row_offset * ROW_SIZE
    }

    pub fn advance(&mut self, num_rows: usize) {
        self.row_num += 1;
        if self.row_num >= num_rows {
            self.end_of_table = true;
        }
    }

    pub fn is_end(&self) -> bool {
        self.end_of_table
    }
}


src/table.rs
use std::process::exit;

use rs_sqlite::{to_u8_array, Row, EXIT_SUCCESS, ROW_SIZE, TABLE_MAX_ROWS};
use scanf::sscanf;

use crate::pager::Pager;

enum ExecuteResult {
    TableFull,
    NoExecute,
    Success,
}
enum PrepareResult {
    Success,
    UnrecognizedStatement,
    SyntaxError,
    ParseStringTooLong,
    ParseNegativeId,
}

#[derive(Debug)]
enum StatementType {
    Select,
    Insert,
    Unknown,
}

#[derive(Debug)]
struct Statement {
    typ: StatementType,
    row: Option<Row>,
}

pub struct Table {
    num_rows: usize,
    pager: Pager,
}

impl Table {
    pub fn db_open(filename: &str) -> Table {
        let pager = Pager::open(filename).unwrap();
        let num_rows = pager.file_length / ROW_SIZE;
        Self { num_rows, pager }
    }

    fn insert(&mut self, statement: &Statement) -> ExecuteResult {
        if self.num_rows >= TABLE_MAX_ROWS {
            return ExecuteResult::TableFull;
        }

        let Some(row) = &statement.row else {
            return ExecuteResult::NoExecute;
        };

        let insert_data = row.encode();
        self.pager.save_row(self.num_rows, insert_data);

        self.num_rows += 1;
        ExecuteResult::Success
    }

    fn select(&mut self, _statement: &Statement) -> ExecuteResult {
        for i in 0..self.num_rows {
            let select_data = self.pager.get_row(i);
            let row = Row::from(&select_data).unwrap();
            println!("{}", row);
        }

        ExecuteResult::Success
    }

    fn execute_statement(&mut self, statement: &Statement) -> ExecuteResult {
        match statement.typ {
            StatementType::Insert => self.insert(statement),
            StatementType::Select => self.select(statement),
            StatementType::Unknown => ExecuteResult::NoExecute,
        }
    }

    pub fn free(&mut self) {
        self.pager.free()
    }

    pub fn run(&mut self, src: &str) {
        if src.starts_with('.') {
            match src {
                ".exit" => {
                    self.free();
                    println!("Bye~");
                    exit(EXIT_SUCCESS);
                }
                _ => {
                    println!("Unrecognized command {}", src);
                }
            }
            return;
        }

        let mut statement = Statement {
            typ: StatementType::Unknown,
            row: None,
        };
        match prepare_statement(src, &mut statement) {
            PrepareResult::Success => match self.execute_statement(&statement) {
                ExecuteResult::Success => {
                    println!("Executed.");
                }
                ExecuteResult::TableFull => {
                    println!("Error: Table full.")
                }
                ExecuteResult::NoExecute => {}
            },
            PrepareResult::UnrecognizedStatement => {
                println!("Unrecognized keyword at start of '{}'.", src);
            }
            PrepareResult::SyntaxError => {
                println!("Syntax error. Could not parse statement.");
            }
            PrepareResult::ParseStringTooLong => {
                println!("String is too long.")
            }
            PrepareResult::ParseNegativeId => {
                println!("ID must be positive.")
            }
        }
    }
}

fn prepare_statement(src: &str, statement: &mut Statement) -> PrepareResult {
    if src.len() < 6 {
        return PrepareResult::UnrecognizedStatement;
    }

    match &src[0..6] {
        "insert" => {
            statement.typ = StatementType::Insert;
            let mut id: i32 = 0;
            let mut username: String = String::new();
            let mut email: String = String::new();
            if sscanf!(src, "insert {} {} {}", id, username, email).is_err() {
                return PrepareResult::SyntaxError;
            }

            if id < 0 {
                return PrepareResult::ParseNegativeId;
            }
            if username.len() > 32 || email.len() > 255 {
                return PrepareResult::ParseStringTooLong;
            }

            statement.row = Some(Row {
                id: id.try_into().unwrap(),
                username: to_u8_array::<32>(&username),
                email: to_u8_array::<255>(&email),
            });
        }
        "select" => statement.typ = StatementType::Select,
        _ => {
            return PrepareResult::UnrecognizedStatement;
        }
    }

    PrepareResult::Success
}
